server/backends/chunkstore/chunkstore.go

```diff
@@ -436,116 +446,130 @@ func (w *ChunkstoreWriter) Close() error {
 	return err
 }
 
+func (w *ChunkstoreWriter) SetContext(ctx context.Context) {
+	w.ctx = ctx
```

Comment:
If I understand this correctly, this modifies the context without acquiring a lock, while the `writeLoop` may be accessing it concurrently. If so, there might be a more thread-safe way to do this.

Is extending the context the only thing you need this for? If so, you might be able to get away with something like this, and simplify some of the PR (I don't know if there are issues with this approach though so I would def get Tyler's opinion as well):

```
ctx = /* ... */
var extendOnce sync.Once
select {
  case <-ctx.Done():
    // Allow a one-time deadline extension for finalization.
   extendOnce.Do(func() { ctx = background.ExtendContextForFinalization(ctx) })
  if isDone(ctx) {
    // timed out
  }
}
```

Reply:
That isn't what it did; `ChunkstoreWriter` was entirely synchronous and only communicated with the writeLoop (the struct used in the goroutine) via the `WriteRequest` and `WriteResult` channels. Regardless, Tyler suggested that we just specify the context for every write, so the model has now shifted to that.